app/Main.hs

```diff
@@ -63,5 +64,45 @@ eval env (App t u) =
             Just value -> eval ((arg, value) : env') expr
         _                            -> Nothing
 
+checkUnused :: Expr -> [Identifier]
+checkUnused expr = go [] expr
+  where
+    go unused (Lit _) = unused
+    go unused (Term id) = delete id unused
+    go env (App expr_l expr_r) = mappend (env \\ removed) added
```

Comment:
if you know that you deal with lists (and not some kind of Monoid), there's no need to be so abstract about concatenation